### PR TITLE
c10::string_view

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26628 [wip] Remove CAFFE_KNOWN_TYPE
* #26619 [wip] caffe2::TypeMeta uses compile time type names
* #26618 [wip] Compile time type names
* **#26616 c10::string_view**

Implement C++17 std::string_view for C++11.

This is useful for compile time type name retrieval which I'm going to stack on top of this.
It is also useful to replace `const std::string&` with throughout our codebase.

Differential Revision: [D17518992](https://our.internmc.facebook.com/intern/diff/D17518992/)

### CODE REVIEW

File: c10/util/string_view.h [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+#include <c10/util/reverse_iterator.h>
+#include <cstring>
+#include <limits>
+#include <stdexcept>
+#include <string>
+
+namespace c10 {
+
+/**
+ * Reimplementation of std::string_view for C++11.
```
Comment (ezyang) [inline c10/util/string_view.h]: Umm... did you implement this yourself?

**Comment by ezyang** [comment]
> retrievaly

What's retrievaly?

**Comment by ezyang** [comment]
I'm not sure I have time to do a thorough review of a reimplementation of `string_view` from scratch...

File: c10/util/string_view.h (hunk 1 above)
Comment (smessmer) [reply c10/util/string_view.h]: yes, I didn't find a public implementation that fit our needs i.e. is constexpr and works with pre-C++17.

**Comment by smessmer** [comment]
This is quite well tested, I probably don't need too thoroughly a review if you don't have the time.

> > retrievaly
> 
> What's retrievaly?

retrieval

**Review by dzhulgakov — APPROVED** [review]
Ok, I read it diagonally and looks plausible. Given extensive testing and that it's mostly constexpr it should be good

File: c10/util/C++17.h [hunk 2]
```diff
@@ -294,6 +294,15 @@ template<class T> inline std::string to_string(T value) {
     return detail::to_string_<T>::call(value);
 }
 
+template <class T>
+constexpr const T& min(const T& a, const T& b) {
```
Comment (gchanan) [inline c10/util/C++17.h]: this is constexpr in C++14, right?  Is there a reason we aren't tracking C++14 / C++17 separately so we can kill the C++14 stuff once we move to that?

**Pull request closed without merging**